### ISSUE TITLE
C23 features aren't supported

### ISSUE THREAD

**Issue by WhoBrokeTheBuild** [microsoft/vscode-cpptools#10696]
### Environment

- OS and Version: Ubuntu 22.04
- VS Code Version: 1.76.2
- C/C++ Extension Version: 1.15.0
- If using SSH remote, specify OS of remote machine: 

### Bug Summary and Steps to Reproduce

Bug Summary:

There was an issue (#10615) that allows you to select c23/gnu23 as your compiler version, but the actual new language features are not implemented yet. Obviously this stuff is bleeding edge, but I struggled trying to get it to work and I wanted to submit an issue.

I would be happy to work on adding these features to the C/C++ language server, but I would need to be pointed in the right direction.

Steps to reproduce:
1. Set your C standard version to c23 or gnu23
2. Open the following code with IntelliSense active
```c
enum TestEnum : short
{
    TEST_ENUM_VALUE,
};

void doSomething() {
    int * test1 = nullptr;
    bool test2 = true; // without #include <stdbool.h>
    enum TestEnum test3 = TEST_ENUM_VALUE;
}
```
3. Verify that these all generate "Problems"
![image](https://user-images.githubusercontent.com/993663/225824093-7c16b397-84a0-41ba-97f9-808e16a9615a.png)

4. Compile with `-std=gnu2x` or equivalent, on a very new compiler (I used clang 17) and verify that there are no errors

Expected behavior:
Generate the same errors that clang 17 would, or equivalent

### Configuration and Logs

```shell
{
    "configurations": [
        {
            "name": "Linux",
            "compilerPath": "/usr/bin/clang-17",
            "cStandard": "c23",
            "compileCommands": "${workspaceFolder}/build/compile_commands.json"
        }
    ],
    "version": 4
}

-------- Diagnostics - 3/17/2023, 1:56:58 AM
Version: 1.15.0
Current Configuration:
{
    "name": "Linux",
    "compilerPath": "/usr/bin/clang-17",
    "cStandard": "c23",
    "compileCommands": "WORKSPACE/build/compile_commands.json",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": false,
    "intelliSenseModeIsExplicit": false,
    "rawCompilerPath": "/usr/bin/clang-17",
    "mergeConfigurations": false,
    "browse": {
        "limitSymbolsToIncludedHeaders": true
    }
}
Translation Unit Mappings:
[ WORKSPACE/src/test.c ]:
    WORKSPACE/src/test.c
Translation Unit Configurations:
[ WORKSPACE/src/test.c ]:
    Process ID: 3852781
    Memory Usage: 544 MB
    Compiler Path: /usr/bin/clang-17
    Includes:
        /usr/lib/llvm-17/lib/clang/17/include
        /usr/local/include
        /usr/include/x86_64-linux-gnu
        /usr/include
    Standard Version: c23
    IntelliSense Mode: linux-clang-x64
    Other Flags:
        --clang
        --clang_version=170000
    compile_commands.json entry:
        directory: WORKSPACE/build
        file: WORKSPACE/src/test.c
        command: /usr/bin/clang   -g -std=gnu2x -o WORKSPACE/build/src/test.c.o -c WORKSPACE/src/test.c
        output: CMakeFiles/slwjson.dir/src/test.c.o
Total Memory Usage: 544 MB
Browse Paths from compile_commands.json, from workspace folder: WORKSPACE
    WORKSPACE/src
```

**Comment by sean-mcmanus** [comment]
The EDG parser used for IntelliSense hasn't implemented the feature yet -- you can check the feature set implemented at https://en.cppreference.com/w/c/23 to track the progress, although it may take some delay before we can ship an update.

**Comment by JPHutchins** [comment]
EDG C23 feature status here: https://docs.google.com/spreadsheets/d/1XGTWYSks6L_20aAi4XiS-3pyi2cS_z_k-NKAroelGX8/edit#gid=0

The chart here, https://en.cppreference.com/w/c/23, is out of date according to that spreadsheet.  For example, the spreadsheet shows support for constexpr object defs and I am not seeing that in VSCode.

**Comment by notpeelz** [comment]
@JPHutchins that google docs link has write permissions.

**Comment by starball5** [comment]
Related on Stack Overflow: [VS Code's C/C++ extension says the C23 true and false keywords are unidentified](https://stackoverflow.com/q/77830499/11107541)

**Comment by markgoetz** [comment]
Is there a update to this?  It appears that per [the Google Doc linked above](https://docs.google.com/spreadsheets/d/1XGTWYSks6L_20aAi4XiS-3pyi2cS_z_k-NKAroelGX8/edit#gid=0), the C2X / EDG parser supports a few things that currently incorrectly throw errors in VSCode - most notably binary literals.

**Comment by bobbrow** [comment]
@markgoetz, what version of the extension are you using?  I forget when we merged EDG 6.5 into our extension, but it's in 1.19.3 as far as I can tell.
![image](https://github.com/microsoft/vscode-cpptools/assets/12818240/9230256f-8b6e-4787-9e85-2561bab52e1d)

EDG 6.6 should be in 1.20.0 and above.

I'm not sure how valuable keeping this bug open is for us though since we're not driving the implementation of individual C23 features, however, it looks like we do need to update our support to pass ms_c23 to EDG, as support for that switch was recently added and we're still clamped to ms_c17.  @browntarik would you be able to prep a PR for that?

**Comment by markgoetz** [comment]
> @markgoetz, which version of the extension are you using? I forget when we merged EDG 6.5 into our extension, but it's in 1.19.3 as far as I can tell. ![image](https://private-user-images.githubusercontent.com/12818240/306353034-9230256f-8b6e-4787-9e85-2561bab52e1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg3OTQ2MzQsIm5iZiI6MTcwODc5NDMzNCwicGF0aCI6Ii8xMjgxODI0MC8zMDYzNTMwMzQtOTIzMDI1NmYtOGI2ZS00Nzg3LTllODUtMjU2MWJhYjUyZTFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI0VDE3MDUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThmNDU1NzU1NGZmMDQyNmRlOTZiYjljMzdlN2IwOTI3NGRhYWZhNTFiMDcxMmNkYmE3YWViNDBmYWJkN2FkZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RHlxAedVxTj0nB5hreNuEVQt2lgijdr4nUpvvGFXBm0)
> 
> EDG 6.6 should be in 1.20.0 and above.
> 
> I'm not sure how valuable keeping this bug open is for us though since we're not driving the implementation of individual C23 features, however, it looks like we do need to update our support to pass ms_c23 to EDG, as support for that switch was recently added and we're still clamped to ms_c17. @browntarik would you be able to prep a PR for that?

I am on 1.19.4 and still seeing the error.

**Comment by bobbrow** [comment]
@markgoetz can you share some code/screenshots and the output of running the `C/C++: Log Diagnostics` command while the file with the error is active in the editor?

**Comment by bobbrow** [comment]
> I noticed "Standard version: c17" for the file I have open. Maybe I have something misconfigured?

Yes, something seems wrong here.  I can confirm on macOS that c17 mode produces the error in your screenshot, and c23 mode does not, so getting the extension to use c23 for your source file seems to be what's blocking you from getting correct IntelliSense here.

The extension will not know how to get information from your `lcc` compiler right now either (#6931 will give you more options when it is complete). For the time being, you may want to set `"compilerPath": ""` to disable the extension's compiler querying attempts.  You may also want to delete `"configurationProvider": "ms-vscode.makefile-tools"` since it will also try to set the compiler to `lcc` (which we don't understand).  Please try those two things and let me know if it gets you any further.  If not, we can try setting `"C_Cpp.loggingLevel": "debug"` and checking your logs.

**Comment by markgoetz** [comment]
I changed the compilerPath and it seems to work!  Thanks for the assistance!

**Comment by christiangda** [comment]
`constexpr` is also not supported

```c
#ifndef PAGE_H
#define PAGE_H

#include <stdlib.h>

constexpr uint32_t PAGE_SIZE = 4096;

typedef struct Page
{
  uint32_t id;
  char data[PAGE_SIZE];
  uint32_t items;
} Page;

#endif // PAGE_H
```

![Screenshot 2024-03-03 at 12 58 42](https://github.com/microsoft/vscode-cpptools/assets/1197820/76766624-dce9-47ce-96e1-a0fb8280052b)

![Screenshot 2024-03-03 at 12 59 19](https://github.com/microsoft/vscode-cpptools/assets/1197820/c0e9fac8-e358-4f61-85f9-f16beafba7c6)

**Comment by 0Alastair1** [comment]
_BitInt also doesn't work

**Comment by rk42745417** [comment]
`nullptr` also doesn't work.

**Comment by bengabay1994** [comment]
This issue is still going on? any ETA for this one maybe?

**Comment by DolphinIQ** [comment]
Any information on this?

**Comment by sean-mcmanus** [comment]
@rk42745417 nullptr hasn't been implemented yet by the EDG parser, see https://en.cppreference.com/w/c/23 (see the 6.5+ table entries)

@bengabay1994 It's still in progress along with c++23. I don't have an ETA.

@DolphinIQ Our latest updates seems have some support for a variety of c23 features like standard attributes, nodiscard attribute, __VA_OPT__, digit separators, `#elifdef`, _Static_assert, etc., although I haven't confirmed all of those (support has been there for over a year).

**Comment by DolphinIQ** [comment]
@vndlgd Digit separators work for me. Make sure you select `c23` as your C standard in extension settings

**Comment by FooqX** [comment]
So when is it going to be fixed? Please fix at least the constexpr, and true/false not being defined.

**Comment by sean-mcmanus** [comment]
@FooqX For the most recent status, you can check the c23 link at https://www.edg.com/c/features , i.e. https://www.edg.com/c23_features.html . It looks like there's a PR created for constexpr/true/false, but they haven't been released yet.

**Comment by podgorskiy** [comment]
@sean-mcmanus,  I would highly appreciate adding support for `constexpr`. Thank you!

**Comment by Raffaello** [comment]
Hi Would it be possible to fit this issue?
is there anything i could do to help?

at least the keywords would it be possible to add them as a quick fix otherwise intellisense will just highlight errors when there are none please?

**Comment by Raffaello** [comment]
even the `true` keyword become an issue with C23 and GCC:

```cpp
// stdbool.h
#if defined __STDC_VERSION__ && __STDC_VERSION__ > 201710L
/* bool, true and false are keywords.  */
#else
#define bool	_Bool
#define true	1
#define false	0
#endif
```

that if define above is returning true, so there is no definition of the true / false and bool defines and VSCode highlight them with errors. very annoying, basically going one step back rather than forward.

**Comment by Raffaello** [comment]
> The EDG parser used for IntelliSense hasn't implemented the feature yet -- you can check the feature set implemented at https://en.cppreference.com/w/c/23 to track the progress, although it may take some delay before we can ship an update.

based on the EDG the nullptr key word has been fixed in version 6.7 but in VSCode is still an error like the bool, true, false keywords.

**Comment by robindegen** [comment]
It's been over 2 years. Any update on this?!

**Comment by wolfield** [comment]
WOW, official extension feels like an abandonware.

**Comment by sean-mcmanus** [comment]
@Raffaello I don't get any error with nullptr with c23. Can you provide more repro info?

**Comment by sean-mcmanus** [comment]
@Raffaello @wolfield @robindegen c23 nullptr support was added in our latest update 1.26.3 (you should not get a squiggle, but it won't color lexically as a keyword since that is implemented by VS Code via https://github.com/jeff-hykin/better-cpp-syntax/issues ).

**Comment by Raffaello** [comment]
@sean-mcmanus 
`nullptr`
<img width="482" height="123" alt="Image" src="https://github.com/user-attachments/assets/598d3f33-d5e4-43c4-8903-dce22924e2c9" />

`bool`

<img width="468" height="116" alt="Image" src="https://github.com/user-attachments/assets/fb3a8a4d-185c-4ae8-9a18-443add3a6d2d" />


`false`

<img width="424" height="112" alt="Image" src="https://github.com/user-attachments/assets/5ea5b62c-91c5-4f63-886e-81a1d20dd385" />


I am using MSYS2 ucrt gcc and the build is succesful but i have the following errors reported by intellisense:

<img width="508" height="137" alt="Image" src="https://github.com/user-attachments/assets/4332d01c-1b42-4a31-a66e-1c7bcf71f682" />


------

a simple repo like this one:
```cmake
#CMakeLists.txt


set(CMAKE_C_STANDARD 23)

add_executable(main main.c)

```

```cpp
// main.c


bool b = false;

int* ptr = nullptr;

int main()
{

    return 0;
}
```

using MSYS2 ucrt GCC (15.1.0)
build succesfully but you should see the mentioned errors:

<img width="504" height="387" alt="Image" src="https://github.com/user-attachments/assets/cb13343e-375b-4f3f-beaa-cbcc3357e611" />

**Comment by sean-mcmanus** [comment]
@Raffaello How is your 1st screenshot different from the last screenshot in which nullptr has no squiggle? The `false` case isn't implemented by our EDG parser yet. I'm seeing no squiggle as in your 2nd/last screenshot.

**Comment by Raffaello** [comment]
> [@Raffaello](https://github.com/Raffaello) How is your 1st screenshot different from the last screenshot in which nullptr has no squiggle? The `false` case isn't implemented by our EDG parser yet. I'm seeing no squiggle as in your 2nd/last screenshot.

none for the value of added information, just as a noise, emotional value: to make it feel that yet little annoying bit... that's also is leading to confusion, a.k.a. not possible to really use vscode in c23 mode yet.

**Comment by aleios** [comment]
EDG? What? Microsoft with your own frontend... why is this held up by like 5 guys in a shack?

**Comment by sean-mcmanus** [comment]
@aleios The Microsoft MSVC cl.exe front-end only supports compiling in msvc mode and not clang/gcc modes, in addition to other factors. Also, the EDG compiler has currently implemented 20 more c23 features than the cl.exe compiler according to https://en.cppreference.com/w/c/compiler_support/23 .

**Comment by aleios** [comment]
@sean-mcmanus 
I know, I know. I'm having a giggle. Sorry

**Comment by sean-mcmanus** [comment]
c23 `bool`, `true`, `false` is available with 1.28.0 pre-release. If anyone has particular c23 features they feel should be prioritized, then let us know via filing a separate issue on it (or upvoting a pre-existing issue...if we closed it as a duplicate of this one we could re-open it).

**Comment by ghost** [comment]
> c23 `bool`, `true`, `false` is available with 1.28.0 pre-release. If anyone has particular c23 features they feel should be prioritized, then let us know via filing a separate issue on it (or upvoting a pre-existing issue...if we closed it as a duplicate of this one we could re-open it).

Personally I feel like constexpr should be priority one.

**Comment by DolphinIQ** [comment]
> c23 `bool`, `true`, `false` is available with 1.28.0 pre-release. If anyone has particular c23 features they feel should be prioritized, then let us know via filing a separate issue on it (or upvoting a pre-existing issue...if we closed it as a duplicate of this one we could re-open it).

I think `constexpr`, enum underlying types, and maybe `#embed` would be greatly appreciated, in that order of importance.

**Comment by sean-mcmanus** [comment]
@DolphinIQ "enum underlying types" was added in https://github.com/microsoft/vscode-cpptools/releases/tag/v1.29.2 . `#embed` is pending relatively soon (`contexpr` is not, unknown ETA).

**Comment by rostidev** [comment]
Hello,

I think I've found the right ticket to report an issue, but if it's not, please tell me an I will open a separate one.

I'm using `ms-vscode.cpptools` version `1.29.3` in VS Code version `1.107.1` on Fedora 43 Linux with GCC version `15.2.1 20251211 (Red Hat 15.2.1-5)`. I play with a small C23 project with `.vscode/c_cpp_properties.json` as following:
```json
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [""],
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "c23",
            "cppStandard": "c++26",
            "intelliSenseMode": "linux-gcc-x64",
            "compilerArgs": ["-std=c23"]
        }
    ],
    "version": 4
}
```
Although it defines `cStandard` as `c23` I have the following  `expected a ';' C/C++(65)` error in my code, after the `constexpr` keyword:

<img width="389" height="111" alt="Image" src="https://github.com/user-attachments/assets/923ae02a-3236-45e3-bbe9-7fff0fcbcbe9" />

But GCC compiles this code without any error, i.e. this is a bug in the `ms-vscode.cpptools` plugin.

Is it a known issue and when it is planned to be resolved?

**Comment by rostidev** [comment]
In addition to my previous message,

If I understood it right the `ms-vscode.cpptools` plugin uses the EDG eccp parser for parsing any C or C++ code, and support of any new C23 keyword like the `constexpr` depends on that parser entirely.

I've found https://www.edg.com/c/features web page that has the following link to the "**C23 features and their implementation status**". Now it redirects to an Excel document in Google Drive:

https://www.edg.com/c23_features.html

The `constexpr` feature of C23 (stored there as `constexpr for Object Definitions`) was recorded in this document in 2022 and we are almost in 2026 now. Why it takes so much time to implement `constexpr` in their C23 parser? Is the source code of this parser available somewhere and open for open-source contributions? It seems like EDG should receive some emergency help from the community, shouldn't it?

**Comment by sean-mcmanus** [comment]
@rostidev We'll know more on the constexpr status/timeline and the open source status of EDG in 2026.

**Comment by rostidev** [comment]
What about empty initializer braces? Seems like `ms-vscode.cpptools` doesn't support them too.

<img width="564" height="109" alt="Image" src="https://github.com/user-attachments/assets/4198d0ed-cea6-4c10-8eac-c0665b7b91b9" />

With `-std=c23` GCC compiles this code without any error or warning.

**Comment by sean-mcmanus** [comment]
@rostidev That is in the same category as constexpr -- it seems like they have a fix in progress for some time next year, but it's not in our pending January update.

**Comment by sean-mcmanus** [comment]
`#embed` IntelliSense support is available in https://github.com/microsoft/vscode-cpptools/releases/tag/v1.30.1 . `constexpr` and empty initializer braces aren't added yet (as mentioned earlier)...we'll let you know when we have any update on those.

**Comment by nabelekt** [comment]
I've just run into this. How is this not a high priority issue?
